### DISCUSSION TITLE
Register new user

### DISCUSSION THREAD

**Discussion by cbirkenbeul** [plausible/analytics#454] [Self-Hosted Support] [answered]
Hi,
is it not possible to register new users on a selfhosted installation? My girlfriend needs here own account for her blogs. If try to register a new account i got a 500 error.
Thanks.

**Comment by ukutaht** [comment]
Hey @cbirkenbeul can you see what error is being reported? You can access the logs by `docker logs <plausible_container_name>`.

**Comment by cbirkenbeul** [comment]
A Webpage with Error 500. But one from you, not from the webserver.

Logs show that its not able to send mail, i guess
```
** (exit) an exception was raised:
    ** (Bamboo.SMTPAdapter.SMTPError) There was a problem sending the email through SMTP.

The error is :retries_exceeded

More detail below:

{:network_failure, 'mail.domain.de', {:error, {:tls_alert, {:unexpected_message, 'TLS client: In state hello at tls_record.erl:477 generated CLIENT ALERT: Fatal - Unexpected Message\n'}}}}

        (plausible 0.0.1) lib/plausible/mailer.ex:14: Plausible.Mailer.send_email/1
        (plausible 0.0.1) lib/plausible_web/controllers/auth_controller.ex:37: PlausibleWeb.AuthController.register/2
        (plausible 0.0.1) lib/plausible_web/controllers/auth_controller.ex:1: PlausibleWeb.AuthController.action/2
        (plausible 0.0.1) lib/plausible_web/controllers/auth_controller.ex:1: PlausibleWeb.AuthController.phoenix_controller_pipeline/2
        (phoenix 1.4.17) lib/phoenix/router.ex:288: Phoenix.Router.__call__/2
        (plausible 0.0.1) lib/plug/error_handler.ex:65: PlausibleWeb.Router.call/2
        (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint.plug_builder_call/2
        (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint."call (overridable 3)"/2
```

TLS cert is valid, i checked it. Its a lets encrypt cert if this matter. 

My mailserver config:

```
MAILER_EMAIL=plausible@domain.de
SMTP_HOST_ADDR=mail.domain.de
SMTP_POST_PORT=587
SMTP_USER_NAME=plausible@domain.de
SMTP_USER_PWD=password
SMTP_HOST_SSL_ENABLED=true
SMTP_RETRIES=20
```

**Reply by ukutaht** [reply]
OK, gotcha.

Plausible is designed so that you need a valid email address to sign up. This means if the app cannot send email, registration does not work at all.

For the cloud instance we actually use Postmark instead of a custom SMTP server. I have very little experience with running your own email server so if you need help fixing email delivery I'm afraid I cannot be too helpful. Maybe try turning off `SMTP_HOST_SSL_ENABLED` if that doesn't pose a security risk in your setup?

You can bypass all of this by creating the user from the app console. To do this:
1. Find your plausible container name with `docker ps`. In my case it was `plausible_hosting_1`
2. Enter the app console `docker exec -it <container-name> /app/bin/plausible remote`
3. Create a user:
```elixir
Plausible.Auth.User.new(%Plausible.Auth.User{}, %{name: "Test name", email: "test@email.com"}) |> Plausible.Auth.User.set_password("123456") |> Plausible.Repo.insert
```

**Comment by cbirkenbeul** [comment]
Hey,
i used a valid email address, because i thought that the software need this. I guess its something with the communication between the container and the mail server. I see very often that some older libraries have problems with let's encrypt certs to communitcate with. Maybe you can have a look for further versions. 
I guess thats also the reason why my weekly or monthly reports are not arrive. 

For this problem i create the user manually, because i dont like to disable my ssl on my webserver. To just skip it on client side is not possible, because i force them on server side. 

Thank you very much for you help, much appreciated.

**Comment by SamGoody** [comment]
When I try the above on version 3.3, I get an error:
```
iex(plausible@1c28bf00d06c)1> Plausible.Auth.User.new(%Plausible.Auth.User{}, %{name: "Rachel", email: "rachel@example.com" }) |> Plausible.Auth.User.set_password("ab_12") |> Plausible.Repo.insert
{:error,
 #Ecto.Changeset<
   action: :insert,
   changes: %{
     email: "rachel@example.com",
     name: "Rachel",
     password: "ab_12",
     password_hash: "$2b$12$24Y0H5HgwhSwoABTeKOd8elmDo3Wdex4AklEeRLi8xPA67GPBvWja",
     trial_expiry_date: ~D[2121-10-10]
   },
   errors: [
     password: {"can't be blank", [validation: :required]},
     password_confirmation: {"can't be blank", [validation: :required]}
   ],
   data: #Plausible.Auth.User<>,
   valid?: false
 >}
```

Are there updated instructions?

**Comment by SamGoody** [comment, marked as answer]
Solved it, just had to read the error code ;/

The set_password method no longer needs to be called separately, instead the password and conform password are passed directly into User.new [in auth/user.ex](https://github.com/plausible/analytics/blob/master/lib/plausible/auth/user.ex)

```
docker exec -it <container-name> /app/bin/plausible remote

Plausible.Auth.User.new(%Plausible.Auth.User{}, \
%{name: "Test name", email: "test@email.com", password: "123456", password_confirmation: "123456" }) \
|> Plausible.Repo.insert
```

It looks like you can also add a list of approved sites, but haven't tried.

Thanks alot to the Plausible team!

**Reply by metmarkosaric** [reply]
glad to hear that! thanks for the update @SamGoody!

**Reply by nchlswhttkr** [reply]
With the release of `v1.4.0` you don't need to pass in the default user struct any more, just the various required attributes.

```
Plausible.Auth.User.new(%{name: "Test name", email: "test@email.com", password: "123456", password_confirmation: "123456" }) |> Plausible.Repo.insert
```

**Reply by ukutaht** [reply]
Also in v1.4.0, the email confirmation behaviour is not enabled by default. If you don't have the mailer set up, you should be able to still register new users through the UI.

**Reply by nsgoyat** [reply]
@nchlswhttkr Thanks for your insights. I added a general user through your answer but how do we add an Admin User to Self-Hosted Plausible using the CLI?
